alchemist/alchemist-fxui/src/main/kotlin/it/unibo/alchemist/wormhole/implementation/LeafletMapWormhole.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2010-2020, Danilo Pianini and contributors
+ * listed in the main project's alchemist/build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.wormhole.implementation
+
+import de.saring.leafletmap.LatLong
+import it.unibo.alchemist.boundary.minus
+import it.unibo.alchemist.boundary.monitors.CustomLeafletMapView
+import it.unibo.alchemist.boundary.plus
+import it.unibo.alchemist.boundary.wormhole.implementation.PointAdapter
+import it.unibo.alchemist.boundary.wormhole.interfaces.Wormhole2D
+import it.unibo.alchemist.model.implementations.positions.LatLongPosition
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.GeoPosition
+import it.unibo.alchemist.runOnFXThread
+import it.unibo.alchemist.syncRunOnFXThread
+import javafx.scene.Node
+import java.awt.Point
+
+/**
+ * The wormhole used for managing a [CustomLeafletMapView].
+ */
+class LeafletMapWormhole(
+    environment: Environment<*, GeoPosition>,
+    node: Node,
+    private val map: CustomLeafletMapView
+) : WormholeFX<GeoPosition>(
+    environment,
+    node
+) {
+    init {
+        mode = Wormhole2D.Mode.MAP
+    }
+
+    override fun getEnvPoint(viewPoint: Point): GeoPosition =
+        syncRunOnFXThread {
+            map.getLatLongFromPoint(viewPoint).toGeoPosition()
+        }
+
+    override fun getViewPoint(envPoint: GeoPosition): Point =
+        syncRunOnFXThread {
+            map.getPointFromLatLong(envPoint.toLatLong())
+        }
+
+    override fun rotateAroundPoint(p: Point?, a: Double) =
+        throw UnsupportedOperationException()
+
+    override fun setEnvPosition(envPoint: GeoPosition) {
+        position = PointAdapter.from(getViewPoint(envPoint))
+        runOnFXThread {
+            map.panTo(envPoint.toLatLong())
+        }
+    }
+
+    override fun setViewPosition(viewPoint: Point) {
+        val movement = viewPosition - viewPoint
+        position = position.sum(PointAdapter.from<GeoPosition>(movement))
+        runOnFXThread {
+            map.panBy(movement.x, movement.y)
+        }
+    }
+
+    override fun optimalZoom() {
+        zoom = CustomLeafletMapView.MAX_ZOOM_VALUE.toDouble()
+        @Suppress("UNCHECKED_CAST")
+        val env = environment as Environment<Any?, GeoPosition>
```

Comment:
Remove suppression. This is unsafe, you should require the type you need... Why Any?

Reply:
The compiler reports that casting `Environment<?, GeoPosition>` to `Environment<Any?, GeoPosition>` is unsafe, which I don't understand. I supposed it was some Java-Kotlin interop issue.

The type of the node of the simulation is not relevant, but without casting it to `Any?` we can't pass elements of `environment.nodes` to `environment.getPosition`, here's the [error](https://i.imgur.com/UOqyOa7.png). Apologies for the image, I couldn't get it formatted nicely in text.

Reply:
The compiler is right: you cant' cast an invarant projected class Environment<?> to an invariant class of a different type, not even if it is the top type (you'd need `Environment` to be covariant in the first parameter).
Anyway... yeah probably little to do here.